sklearn/base.py

```diff
@@ -250,6 +252,30 @@ def get_params(self, deep=True):
                 deep_items = value.get_params().items()
                 out.update((key + "__" + k, val) for k, val in deep_items)
             out[key] = value
+
+        init_func = getattr(self.__init__, "deprecated_original", self.__init__)
+        init_params = inspect.signature(init_func).parameters
+        init_params = {name: param.default for name, param in init_params.items()}
+
+        def has_changed(k, v):
+            if k not in init_params:  # happens if k is part of a **kwargs
+                return True
+            if init_params[k] == inspect._empty:  # k has no default value
+                return True
+            # try to avoid calling repr on nested estimators
+            if isinstance(v, BaseEstimator) and v.__class__ != init_params[k].__class__:
+                return True
+            # Use repr as a last resort. It may be expensive.
+            if repr(v) != repr(init_params[k]) and not (
+                is_scalar_nan(init_params[k]) and is_scalar_nan(v)
+            ):
+                return True
+            return False
+
+        non_default = list({k for k, v in out.items() if has_changed(k, v)})
+        out = ParamsDict(out)
+        out.non_default = non_default
+
```

Comment:
Lines 260 to 273 taken from `utils._pprint._changed_params`. I wasn't able to just call that function because it also calls `get_params()` - It gets into an infinite loop.